pkg/kclient/deployments.go

```diff
@@ -210,3 +253,77 @@ func (c *Client) DeleteDynamicResource(name, group, version, resource string) er
 
 	return c.DynamicClient.Resource(deploymentRes).Namespace(c.Namespace).Delete(name, &metav1.DeleteOptions{})
 }
+
+// Define a function that is meant to create patch based on the contents of the deployment
+type deploymentPatchProvider func(deployment *appsv1.Deployment) (string, error)
+
+// LinkSecret links a secret to the Deployment of a component
+func (c *Client) LinkSecret(secretName, componentName, applicationName string) error {
+
+	var deploymentPatchProvider = func(d *appsv1.Deployment) (string, error) {
+		if len(d.Spec.Template.Spec.Containers[0].EnvFrom) > 0 {
+			// we always add the link as the first value in the envFrom array. That way we don't need to know the existing value
+			return fmt.Sprintf(`[{ "op": "add", "path": "/spec/template/spec/containers/0/envFrom/0", "value": {"secretRef": {"name": "%s"}} }]`, secretName), nil
+		}
+
+		//in this case we need to add the full envFrom value
+		return fmt.Sprintf(`[{ "op": "add", "path": "/spec/template/spec/containers/0/envFrom", "value": [{"secretRef": {"name": "%s"}}] }]`, secretName), nil
+	}
+
+	return c.patchDeploymentOfComponent(componentName, applicationName, deploymentPatchProvider)
+}
+
+// UnlinkSecret unlinks a secret to the Deployment of a component
+func (c *Client) UnlinkSecret(secretName, componentName, applicationName string) error {
+	// Remove the Secret from the container
+	var deploymentPatchProvider = func(d *appsv1.Deployment) (string, error) {
+		indexForRemoval := -1
+		for i, env := range d.Spec.Template.Spec.Containers[0].EnvFrom {
+			if env.SecretRef.Name == secretName {
+				indexForRemoval = i
+				break
+			}
+		}
+
+		if indexForRemoval == -1 {
+			return "", fmt.Errorf("Deployment does not contain a link to %s", secretName)
+		}
+
+		return fmt.Sprintf(`[{"op": "remove", "path": "/spec/template/spec/containers/0/envFrom/%d"}]`, indexForRemoval), nil
+	}
+
+	return c.patchDeploymentOfComponent(componentName, applicationName, deploymentPatchProvider)
+}
+
+// this function will look up the appropriate DC, and execute the specified patch
```

Comment:
Comments should always start with the function name `patchDeploymentOfComponent` or else `gofmt` will complain / get mad